execution/evm/state.go

```diff
@@ -194,7 +215,60 @@ func (st *State) InitCode(address crypto.Address, code []byte) {
 			"tried to initialise code for a contract that already exists: %v", address))
 		return
 	}
+
 	acc.EVMCode = code
+
+	// keccak256 hash of a contract's code
+	hash := sha3.NewKeccak256()
+	hash.Write(code)
+	codehash := hash.Sum(nil)
+
+	forebear := &address
+	metamap := acc.MetaMap
+	if parent != nil {
+		// find our ancestor, i.e. the initial contract that was deployed, from which this contract descends
+		ancestor := st.mustAccount(*parent)
+		if ancestor.Forebear != nil {
+			ancestor = st.mustAccount(*ancestor.Forebear)
+			forebear = ancestor.Forebear
+		} else {
+			forebear = parent
+		}
+		metamap = ancestor.MetaMap
+	}
+
+	// If we have a list of ABIs for this contract, we also know what contract code it is allowed to create
+	// For compatibility with older contracts, allow any contract to be created if we have no mappings
+	if metamap != nil && len(metamap) > 0 {
+		found := false
+		for _, m := range metamap {
+			if bytes.Equal(codehash, m.CodeHash) {
+				found = true
+				break
```

Comment:
Personally I'd rather see this broken out into a function where we can replace the `break`s with `return`s... I guess it's a matter of taste but I find that easier to follow with nesting etc.

Reply:
Good point. The code is shorter now!